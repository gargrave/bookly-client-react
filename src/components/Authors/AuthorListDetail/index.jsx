import React from 'react'
import { func, number, shape, string } from 'prop-types'

import './styles.css'

const AuthorListDetail = (props) => {
  return (
    <div className="author-list-detail" onClick={props.onClick}>
      <p className="author-name">
        {props.author.firstName} {props.author.lastName}
      </p>
    </div>
  )
}

AuthorListDetail.propTypes = {
  author: shape({
    id: number.isRequired,
    firstName: string.isRequired,
    lastName: string.isRequired,
    createdAt: string.isRequired,
    updatedAt: string.isRequired,
  }),
  onClick: func.isRequired,
}

export default AuthorListDetail
