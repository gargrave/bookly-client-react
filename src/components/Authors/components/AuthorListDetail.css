div.author-list-detail {
  cursor: pointer;
  border: solid 1px #bbb;
  border-radius: 5px;
  box-shadow: 0 1px 3px rgba(0, 0, 0, .12), 0 1px 2px rgba(0, 0, 0, .24);
  margin: 10px auto;
  max-width: 600px;
  padding: 8px;
}

div.author-list-detail:hover {
  background: #ffe8e4;
}

div.author-list-detail p {
  margin: 0;
}
