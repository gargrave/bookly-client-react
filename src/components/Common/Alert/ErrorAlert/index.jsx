import React from 'react'
import { string } from 'prop-types'

import { prependClass } from '@/utils/cssHelpers'

const classes = () => {
  return `${prependClass('alert')} alert alert-danger`
}

const ErrorAlert = ({
  error,
}) => (
  <div className={classes()}>
    {error}
  </div>
)

ErrorAlert.propTypes = {
  error: string.isRequired,
}

export default ErrorAlert
