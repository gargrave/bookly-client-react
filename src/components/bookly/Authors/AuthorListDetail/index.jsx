// @flow
import React from 'react';
import { func, shape, string } from 'prop-types';

import type { Author } from '../../../../constants/flowtypes';

import Card from '../../../common/Card';

type Props = {
  author: Author,
  onClick: Function,
};

function AuthorListDetail({
  author,
  onClick,
}: Props) {
  return (
    <Card
      classes={['author-card']}
      onClick={onClick}
      title={`${author.firstName} ${author.lastName}`}
    />
  );
}

AuthorListDetail.propTypes = {
  author: shape({
    firstName: string.isRequired,
    lastName: string.isRequired,
  }),
  onClick: func.isRequired,
};

export default AuthorListDetail;
