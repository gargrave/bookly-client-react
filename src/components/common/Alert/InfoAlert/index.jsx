import React from 'react'
import { string } from 'prop-types'

import { prependClass } from '@/utils/cssHelpers'

const classes = () => {
  return `${prependClass('alert')} alert alert-info`
}

const InfoAlert = ({
  message,
}) => (
  <div className={classes()}>
    {message}
  </div>
)

InfoAlert.propTypes = {
  message: string.isRequired,
}

export default InfoAlert
