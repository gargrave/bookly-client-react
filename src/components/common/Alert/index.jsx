import React from 'react'
import { oneOf, string } from 'prop-types'

import { buildClasses } from '@/utils/cssHelpers'

const Alert = ({
  message,
  type,
}) => (
  <div className={buildClasses('alert', ['alert', `alert-${type}`])}>
    {message}
  </div>
)

Alert.propTypes = {
  message: string.isRequired,
  type: oneOf([
    'primary',
    'secondary',
    'success',
    'info',
    'warning',
    'danger',
    'light',
    'dark',
  ]).isRequired,
}

export default Alert
