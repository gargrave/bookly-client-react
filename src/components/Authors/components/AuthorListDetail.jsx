import React from 'react'
import { func, number, shape, string } from 'prop-types'

import './AuthorListDetail.css'

const AuthorListDetail = props => {
  return (
    <div className="author-list-detail" onClick={props.onClick}>
      <p>
        {props.author.firstName} {props.author.lastName}
      </p>
    </div>
  )
}

AuthorListDetail.propTypes = {
  author: shape({
    id: number.isRequired,
    firstName: string.isRequired,
    lastName: string.isRequired,
    createdAt: string.isRequired,
    updatedAt: string.isRequired
  }),
  onClick: func.isRequired
}

export default AuthorListDetail
