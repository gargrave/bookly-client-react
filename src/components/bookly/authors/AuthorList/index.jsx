// @flow
import React from 'react';
import { array, func, string } from 'prop-types';

import type { Author } from '../../../../constants/flowtypes';

import Alert from '../../../common/Alert/';
import AuthorListDetail from '../AuthorListDetail';

type Props = {
  authors: Author[],
  filterBy?: string,
  onAuthorClick: Function,
};

function filterAuthor(author: Author, filterBy?: string = '') {
  if (!filterBy) {
    return true;
  }
  const name = `${author.firstName} ${author.lastName}`.toLowerCase();
  return name.includes(filterBy);
}

function authorList(
  authors: Author[],
  onAuthorClick: Function,
  filterBy?: string,
) {
  return (
    <div>
      {authors
        .filter((a: Author) => filterAuthor(a, filterBy))
        .map((author) =>
          <AuthorListDetail
            author={author}
            key={author.id}
            onClick={onAuthorClick.bind(null, author.id)}
          />
        )
      }
    </div>
  );
}

function noAuthorsMessage() {
  return (
    <Alert
      message={'No Authors created yet!'}
      type={'info'}
    />
  );
}

function AuthorList({
  authors,
  filterBy,
  onAuthorClick,
}: Props) {
  return (
    authors.length
      ? authorList(authors, onAuthorClick, filterBy)
      : noAuthorsMessage()
  );
}

AuthorList.propTypes = {
  authors: array.isRequired,
  filterBy: string,
  onAuthorClick: func.isRequired,
};

export default AuthorList;
