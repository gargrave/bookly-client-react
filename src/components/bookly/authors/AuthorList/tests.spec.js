import React from 'react';
import { mount, shallow } from 'enzyme';

import Alert from '../../../common/Alert';
import AuthorList from './';
import AuthorListDetail from '../../../bookly/authors/AuthorListDetail';

describe('AuthorList', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = {
      authors: [
        { id: 0, firstName: 'A', lastName: 'B' },
        { id: 1, firstName: 'C', lastName: 'D' },
        { id: 2, firstName: 'E', lastName: 'F' },
      ],
      onAuthorClick: jest.fn(),
    };
  });

  it('matches the snapshot', () => {
    wrapper = shallow(<AuthorList {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the correct number of AuthorListDetail wrappers', () => {
    wrapper = mount(<AuthorList {...props} />);
    expect(wrapper.find(AuthorListDetail)).toHaveLength(props.authors.length);
    expect(wrapper.find(Alert)).toHaveLength(0);
  });

  it('renders a nice message when there are no authors', () => {
    props.authors = [];
    wrapper = mount(<AuthorList {...props} />);
    expect(wrapper.find(AuthorListDetail)).toHaveLength(0);
    expect(wrapper.find(Alert)).toHaveLength(1);
  });
});
