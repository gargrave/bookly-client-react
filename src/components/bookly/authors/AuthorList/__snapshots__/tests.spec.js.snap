// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorList matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AuthorList
    authors={
      Array [
        Object {
          "firstName": "A",
          "id": 0,
          "lastName": "B",
        },
        Object {
          "firstName": "C",
          "id": 1,
          "lastName": "D",
        },
        Object {
          "firstName": "E",
          "id": 2,
          "lastName": "F",
        },
      ]
    }
    onAuthorClick={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <AuthorListDetail
          author={
            Object {
              "firstName": "A",
              "id": 0,
              "lastName": "B",
            }
          }
          onClick={[Function]}
        />,
        <AuthorListDetail
          author={
            Object {
              "firstName": "C",
              "id": 1,
              "lastName": "D",
            }
          }
          onClick={[Function]}
        />,
        <AuthorListDetail
          author={
            Object {
              "firstName": "E",
              "id": 2,
              "lastName": "F",
            }
          }
          onClick={[Function]}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "function",
        "props": Object {
          "author": Object {
            "firstName": "A",
            "id": 0,
            "lastName": "B",
          },
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "author": Object {
            "firstName": "C",
            "id": 1,
            "lastName": "D",
          },
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "author": Object {
            "firstName": "E",
            "id": 2,
            "lastName": "F",
          },
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <AuthorListDetail
            author={
              Object {
                "firstName": "A",
                "id": 0,
                "lastName": "B",
              }
            }
            onClick={[Function]}
          />,
          <AuthorListDetail
            author={
              Object {
                "firstName": "C",
                "id": 1,
                "lastName": "D",
              }
            }
            onClick={[Function]}
          />,
          <AuthorListDetail
            author={
              Object {
                "firstName": "E",
                "id": 2,
                "lastName": "F",
              }
            }
            onClick={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "function",
          "props": Object {
            "author": Object {
              "firstName": "A",
              "id": 0,
              "lastName": "B",
            },
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "author": Object {
              "firstName": "C",
              "id": 1,
              "lastName": "D",
            },
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "author": Object {
              "firstName": "E",
              "id": 2,
              "lastName": "F",
            },
            "onClick": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
